sqlmesh/core/macros.py

```diff
@@ -369,10 +370,31 @@ def add_one(evaluator: MacroEvaluator, column: exp.Literal) -> exp.Add:
 
     registry_name = "macros"
 
+    def __init__(self, name: str = "", coerce_args: bool = False) -> None:
+        super().__init__(name)
+        self.coerce_args = coerce_args
+
     def __call__(
-        self, func: t.Callable[..., DECORATOR_RETURN_TYPE]
+        self,
+        func: t.Callable[..., DECORATOR_RETURN_TYPE],
     ) -> t.Callable[..., DECORATOR_RETURN_TYPE]:
-        wrapper = super().__call__(func)
+        @wraps(func)
+        def _typed_func(*args: t.Any, **kwargs: t.Any) -> DECORATOR_RETURN_TYPE:
+            spec = inspect.getfullargspec(func)
+            kwargs = inspect.getcallargs(func, *args, **kwargs)
+            for param, value in kwargs.items():
```

Comment:
what happens if there are no types in the macro definition?

Reply:
Then `spec.annotations` is an empty dict
<img width="1556" alt="image" src="https://github.com/TobikoData/sqlmesh/assets/41213451/2beb52d0-fc65-404f-8368-9137f1114fd7">

We would see it manifest as a KeyError currently but we can catch that and reraise with something more informative, IE all types must be annotated when `coerce_args=True` or something.

Reply:
no, we shouldn't require types. it should be an optional thing, so make sure we have a test for that and add code to handle untyped definitions

Reply:
what do you think about not having coercable_types at all, and just doing it if you add type annotations?

Reply:
If we make the types optional, meaning no annotation just passes it through as-is to the function body -- then yes we can technically just drop coercible_args since you can opt out by not typing it in the sig and doing it in the body manually. 

I can do that

Reply:
@tobymao done